### PR TITLE
refactor and improve merge tests, big time

### PR DESCRIPTION
**Pull request by ajtmccarty**
as a precursor to completely rewriting the merge logic (currently in progress) this PR refactors the merge-related component tests and makes them much more thorough

## Refactoring

the merge tests were mostly in `test_diff_and_merge.py`, which was ~3,000 lines. the tests from that file have been split up and moved into different files within the new `backend/tests/component/core/diff/merge` directory. some of the tests from that file have been replaced by the much more thorough tests within the new testing matrix described below.

## New testing matrix

the previous tests in `test_diff_and_merge.py` covered individual cases in a piecemeal manner and left gaps for bugs to hide in. we had some tests that covered handling schemas with migrated kinds/inheritance and some that covered adding or removing specific data on the user or default branch, but these different scenarios were not tested in a comprehensive manner. the new tests in `test_clean_merge`, `test_conflict_resolved_[base|branch]`, and `test_[source|branch]_target_migration` combine all of the changes that can exist on a diff with conflicts and node kind migrations to provide much more robust coverage of the merge logic.

the new coverage is so good, in fact, that five of the tests have been marked as `xfail`ing until my PR to refactor the merge logic is ready.

### CODE REVIEW

File: backend/tests/component/core/diff/merge/test_agnostic_node_merge.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_agnostic_node_merge.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_branch_local_scope.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_branch_local_scope.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_clean_merge.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_clean_merge.py:1]: part of the new testing matrix

File: backend/tests/component/core/diff/merge/test_conflict_resolved_base.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_conflict_resolved_base.py:1]: new testing matrix

File: backend/tests/component/core/diff/merge/test_conflict_resolved_source.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_conflict_resolved_source.py:1]: new testing matrix

File: backend/tests/component/core/diff/merge/test_delete_edge_cases.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_delete_edge_cases.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_hierarchy.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_hierarchy.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_manual_conflict_resolution.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_manual_conflict_resolution.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_mixed_conflict_peer.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_mixed_conflict_peer.py:1]: new standalone test

File: backend/tests/component/core/diff/merge/test_multi_migration.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_multi_migration.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_schema_default_values.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_schema_default_values.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_sequential_rel_property_updates.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_sequential_rel_property_updates.py:1]: moved from `test_diff_and_merge.py`

File: backend/tests/component/core/diff/merge/test_source_branch_migration.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_source_branch_migration.py:1]: new testing matrix

File: backend/tests/component/core/diff/merge/test_target_branch_migration.py
Comment (ajtmccarty) [inline backend/tests/component/core/diff/merge/test_target_branch_migration.py:1]: new testing matrix

File: backend/tests/component/core/diff/merge/_setup.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+    relationship_name: str,
+    existing_peer_id: str,
+    branch_user: str,
+) -> DeletedRelationshipCtx:
+    """Clear an optional one-cardinality relationship (set to None)."""
+    original_updated_at, original_updated_by = await _fetch_rel_timestamps(
+        db=db, node_id=node_on_main.id, relationship_name=relationship_name, peer_id=existing_peer_id
+    )
+    branch_node = await NodeManager.get_one(db=db, branch=branch, id=node_on_main.id)
+    await branch_node.get_relationship(relationship_name).update(db=db, data=None)
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_setup.py:206]: No important or blocking
From a typing perspective, maybe `data=[None]` would be better. In case one day we remove the check inside `Node.update()` which encapsulate in a list if required.

File: backend/tests/component/core/diff/merge/_setup.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+    # Capture original property value from a fresh read of the main-branch node.
+    # Source/owner need LINKED_NODES and attribute timestamps need USER_TIMESTAMPS.
+    main_node = await NodeManager.get_one(
+        db=db,
+        id=node_on_main.id,
+        include_metadata=MetadataOptions.LINKED_NODES | MetadataOptions.USER_TIMESTAMPS,
+    )
+    main_attr = main_node.get_attribute(attribute_name)
+    original_peer_id = (
+        main_attr.source_id if property_name == "source" else (main_attr.owner_id if property_name == "owner" else None)
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_setup.py:285]: Typing of `property_name` should be a `Literal["source", "owner", "is_protected"]` don't you agree? Or is this referencing a production code logic which is not encapsulated in a `Literal` typing? 
Either way,`StrEnum` seems over-engineered in such case.

File: backend/tests/component/core/diff/merge/_setup.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+    )
+    original_bool = main_attr.is_protected if property_name == "is_protected" else None
+
+    branch_node = await NodeManager.get_one(db=db, branch=branch, id=node_on_main.id)
+    attr = branch_node.get_attribute(attribute_name)
+    if property_name in ("source", "owner"):
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_setup.py:291]: There it would be `Literal["source", "owner"]`. There are a few cases like this one in the file.

File: backend/tests/component/core/diff/merge/_setup.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+    *,
+    node_on_main: Node,
+    relationship_name: str,
+    peer_id: str,
+    property_name: str,
+    branch_user: str,
+) -> ClearedRelationshipPropertyCtx:
+    """Clear a source/owner on an existing relationship.
+
+    Calls ``clear_source()`` / ``clear_owner()`` on the Relationship instance
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_setup.py]: These comments should be removed from the docstring, as they are implementation details. If they are required, perhaps they should be included in the method body comments?

File: backend/tests/component/core/diff/merge/_validators.py [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+    assert attr._get_updated_by() == ctx.branch_user
+
+
+async def validate_updated_relationship_property(
+    db: InfrahubDatabase,
+    branch: Branch,
+    ctx: UpdatedRelationshipPropertyCtx,
+    merge_at: Timestamp,
+) -> None:
+    # NodeManager fetch with prefetch_relationships=True populates relationship
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_validators.py]: Small question regarding this: do you know whether this behavior is expected from the system or not?

File: backend/tests/component/core/diff/merge/_conflict_setup.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        ] = conflict_peer_node.id
+        conflicts.cleared_relationship_property_base_user = base_user
+
+    return conflicts
+
+
+def _conflicting_value(current: Any, original: Any) -> Any:
+    """Pick a value that differs from both ``current`` (branch-side) and ``original``."""
+    if isinstance(current, bool) or isinstance(original, bool):
+        return not bool(current)
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_conflict_setup.py]: There are cases where the value will not be different from `current` and `original` right? Is this a problem?

**Review by polmichel — COMMENTED** [review]
I have juste done `_contexts.py`, `_setup.py`, `_validators.py`, `_matrix_setup.py`, `conflict_setup.py`, `_migrations.py` for now. No blocking comments, I have just not finished the review yet. Still posting the few optional comments.

File: backend/tests/component/core/diff/merge/test_clean_merge.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+from ._validators import validate_all_applied, validate_all_rolled_back
+from .conftest import get_diff_coordinator, get_diff_merger
+
+if TYPE_CHECKING:
+    from infrahub.core.branch import Branch
+    from infrahub.core.node import Node
+    from infrahub.database import InfrahubDatabase
+
+
+async def test_clean_merge_covers_all_change_types(
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/test_clean_merge.py:27]: This is clean.

File: backend/tests/component/core/diff/merge/_conflict_setup.py (hunk 6 above)
Comment (ajtmccarty) [reply backend/tests/component/core/diff/merge/_conflict_setup.py]: true, but it would only be for Booleans and those are not actually used in this code path. adding an error if a Boolean is submitted

File: backend/tests/component/core/diff/merge/_setup.py (hunk 4 above)
Comment (ajtmccarty) [reply backend/tests/component/core/diff/merge/_setup.py]: definitely right. removed

File: backend/tests/component/core/diff/merge/_setup.py (hunk 2 above)
Comment (ajtmccarty) [reply backend/tests/component/core/diff/merge/_setup.py:285]: good point. added the `Literal` typing

File: backend/tests/component/core/diff/merge/_validators.py (hunk 5 above)
Comment (ajtmccarty) [reply backend/tests/component/core/diff/merge/_validators.py]: I don't know if "expected" is the right word. it's a problem that I didn't have time to address when working on the metadata and it is a completely backend issue that the user can't encounter. 

that said, it is annoying and will cause a problem some day, so looking at a fix now

Comment (ajtmccarty) [reply backend/tests/component/core/diff/merge/_validators.py]: added a PR for it here #9054

File: backend/tests/functional/api/test_load_schema.py [hunk 8]
```diff
@@ -596,3 +596,170 @@ async def test_remove_optional_text_field_value(
...
+        # Step 2: switch generic identity to new_id
+        switch = await client.schema.load(schemas=[_BUG5B_SWITCH_GENERIC_SCHEMA])
+        assert not switch.errors
+        assert switch.schema_updated
+
+        parent_schema = await client.schema.get(kind="Bug5bParent", refresh=True)
+        child_schema = await client.schema.get(kind="Bug5bChild", refresh=True)
+        assert parent_schema.human_friendly_id == ["new_id__value"]
+        # Identity fields do not cascade today — child still references old_id.
+        # This is the inconsistent state that triggers the bug in step 3.
```
Comment (polmichel) [inline backend/tests/functional/api/test_load_schema.py]: For instance, maybe something like: _Child should reference the new ID, since identity fields should cascade from the parent schema._

File: backend/tests/functional/api/test_load_schema.py [hunk 9]
```diff
@@ -596,3 +596,170 @@ async def test_remove_optional_text_field_value(
...
+      1. Baseline: generic Bug5bParent with attrs old_id + new_id, hfid=[old_id__value].
+         Node Bug5bChild inherits from Bug5bParent.
+         Both generic and child end up with uniqueness_constraints=[[old_id__value]].
+      2. Switch generic hfid to [new_id__value]. Identity fields do not cascade,
+         so Bug5bChild still carries uniqueness_constraints=[[old_id__value]] — this
+         is the inconsistent state that the bug is about.
+      3. Remove old_id from the generic with state: absent.
+         Today this raises HTTP 500 from the post-write schema.process() in
+         sync_uniqueness_constraints_and_unique_attributes, and the write partially
+         commits, wedging the branch.
```
Comment (polmichel) [inline backend/tests/functional/api/test_load_schema.py]: Still an optional comment: in my opinion, tests should not be about Jira issues or the bugs they are protecting against. It would be easier if all tests described in their docstring what the expected behavior of the system should be. While some edge cases may require explicit specification of what is expected in the test, I believe that an assertion or comment at assertion level is usually sufficient.

In this regard, these texts could be modified or removed
_Identity fields do not cascade, so Bug5bChild still carries uniqueness_constraints=[[old_id__value]] — this is the inconsistent state that the bug is about._

&

_Today this raises HTTP 500 from the post-write schema.process() in sync_uniqueness_constraints_and_unique_attributes, and the write partially commits, wedging the branch._

File: backend/tests/functional/api/test_load_schema.py [hunk 10]
```diff
@@ -596,3 +596,170 @@ async def test_remove_optional_text_field_value(
...
+        # Identity fields do not cascade today — child still references old_id.
+        # This is the inconsistent state that triggers the bug in step 3.
+        assert child_schema.uniqueness_constraints == [["old_id__value"]]
+
+        # Step 3: remove old_id from the generic. This is what blows up today —
+        # the post-write schema.process() reruns sync_uniqueness_constraints_and_
+        # unique_attributes, which fails because the child's uniqueness_constraints
+        # still reference old_id__value. The failure itself is acceptable — the
+        # bug is that the failed update is not rolled back, leaving the branch
+        # with old_id deleted from the generic but still referenced by the child.
```
Comment (polmichel) [inline backend/tests/functional/api/test_load_schema.py]: _Step 3: Remove old_id from the generic schema. Either the schema loading should fail, triggering the rollback process, or it should succeed._

File: backend/tests/functional/api/test_load_schema.py [hunk 11]
```diff
@@ -596,3 +596,170 @@ async def test_remove_optional_text_field_value(
...
+            # If the server returned a structured response, it must be either a
+            # clean 4xx (surfaced via .errors) or a success that cascaded.
+            remove_failed = bool(remove.errors)
+
+        # Whether the load failed loudly (exception / 5xx) or cleanly (.errors),
+        # if it did not succeed then the schema must look exactly like it did
+        # before step 3 — old_id still present on the generic, and the child's
+        # stale uniqueness_constraints untouched. This is what's broken today:
+        # the attribute delete commits to the DB even though schema.process()
+        # raised, leaving the branch in an unrecoverable state.
```
Comment (polmichel) [inline backend/tests/functional/api/test_load_schema.py]: I feel this comment can be entirely removed

File: backend/tests/functional/api/test_load_schema.py [hunk 12]
```diff
@@ -596,3 +596,170 @@ async def test_remove_optional_text_field_value(
...
+            # updated to no longer reference the removed attribute.
+            parent_schema = await client.schema.get(kind="Bug5bParent", refresh=True)
+            child_schema = await client.schema.get(kind="Bug5bChild", refresh=True)
+            assert "old_id" not in {a.name for a in parent_schema.attributes}
+            for constraint_paths in child_schema.uniqueness_constraints or []:
+                assert "old_id__value" not in constraint_paths, (
+                    f"Child still references removed attribute: {constraint_paths}"
+                )
+
+        # Regardless of which outcome we took, the branch must not be wedged —
```
Comment (polmichel) [inline backend/tests/functional/api/test_load_schema.py]: This can also be removed, since the following assertions include an error message in case of failure.

File: backend/tests/functional/api/test_load_schema.py [hunk 13]
```diff
@@ -596,3 +596,170 @@ async def test_remove_optional_text_field_value(
...
+            "namespace": "Bug5b",
+            "attributes": [
+                {"name": "old_id", "kind": "Number", "state": "absent"},
+            ],
+        }
+    ],
+}
+
+
+class TestLoadSchemaInheritedIdentityAttributeRemoval(TestInfrahubApp):
```
Comment (polmichel) [inline backend/tests/functional/api/test_load_schema.py]: Can this test really result in successful schema loading when the `remove_failed` value is false? If so, this seems like a flaky test that we should stabilize. Then, we should create a dedicated schema loading succeeding test to reproduce the succeeding test case that this test is partly about.

File: backend/tests/component/core/diff/merge/_setup.py [hunk 14]
```diff
@@ -0,0 +1,480 @@
...
+    relationship_name: str,
+    existing_peer_id: str,
+    branch_user: str,
+) -> DeletedRelationshipCtx:
+    """Clear an optional one-cardinality relationship (set to None)."""
+    original_updated_at, original_updated_by = await _fetch_rel_timestamps(
+        db=db, node_id=node_on_main.id, relationship_name=relationship_name, peer_id=existing_peer_id
+    )
+    branch_node = await NodeManager.get_one(db=db, branch=branch, id=node_on_main.id)
+    await branch_node.get_relationship(relationship_name).update(db=db, data=None)
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_setup.py:206]: I'm just raising this one again in case you missed it. However, this is still neither important nor blocking. https://github.com/opsmill/infrahub/pull/9042#discussion_r3137117899

File: backend/tests/component/core/diff/merge/_validators.py [hunk 15]
```diff
@@ -0,0 +1,698 @@
...
+
+
+async def validate_all_applied_with_conflict_to_base(  # noqa: C901
+    db: InfrahubDatabase,
+    branch: Branch,
+    contexts: MatrixContexts,
+    base_conflicts: BaseConflicts,
+    merge_at: Timestamp,
+    *,
+    added_node_state: str = "applied",
```
Comment (polmichel) [inline backend/tests/component/core/diff/merge/_validators.py]: Looks like a `Literal["applied", "missing"]`

**Review by polmichel — APPROVED** [review]
This is a great test suite that will inspire confidence in the system current behavior and future refactoring.
Optional comments have been created. The only thing I would care about is the functional test that could end in failure or success. I'm not sure if this kind of behavior should be included in our test suite.

File: backend/tests/component/core/diff/merge/_validators.py (hunk 5 above)
Comment (polmichel) [reply backend/tests/component/core/diff/merge/_validators.py]: Approved !

File: backend/tests/functional/api/test_load_schema.py (hunk 9 above)
Comment (ajtmccarty) [reply backend/tests/functional/api/test_load_schema.py]: yes this is my bad. this whole test was not meant to be included. it was for troubleshooting a different issue. I've deleted it

File: backend/tests/component/core/diff/merge/_validators.py (hunk 15 above)
Comment (ajtmccarty) [reply backend/tests/component/core/diff/merge/_validators.py]: good point. update to a `Literal`
